Fix empty secret namespace issue in reconciler

zakisk opened the pull request:
Fixed empty secret namespace issue in reconcile when git proivder is not github. added E2E tests for gitea as well to confirm this behavior.

## What's happening?
Let's take a scenario that I got 5 PipelineRun and I set `concurrency_limit` to 3 so when PRs are acquired in [queue_pipelinerun](https://github.com/openshift-pipelines/pipelines-as-code/blob/51737746e4914b24de048e36d3ac7e43f16155f5/pkg/reconciler/queue_pipelineruns.go#L54) from semaphore (I got 3 PRs in [acquired list](https://github.com/openshift-pipelines/pipelines-as-code/blob/51737746e4914b24de048e36d3ac7e43f16155f5/pkg/reconciler/queue_pipelineruns.go#L54)) and [looped](https://github.com/openshift-pipelines/pipelines-as-code/blob/51737746e4914b24de048e36d3ac7e43f16155f5/pkg/reconciler/queue_pipelineruns.go#L59), for first PipelineRun [`r.secretNS`](https://github.com/openshift-pipelines/pipelines-as-code/blob/51737746e4914b24de048e36d3ac7e43f16155f5/pkg/reconciler/reconciler.go#L45) value is empty in the loop [at this point](https://github.com/openshift-pipelines/pipelines-as-code/blob/51737746e4914b24de048e36d3ac7e43f16155f5/pkg/reconciler/queue_pipelineruns.go#L66) when calling [updatePipelineRunToInProgress](https://github.com/openshift-pipelines/pipelines-as-code/blob/51737746e4914b24de048e36d3ac7e43f16155f5/pkg/reconciler/reconciler.go#L221) and updatePipelineRunToInProgress returning error from [this line](https://github.com/openshift-pipelines/pipelines-as-code/blob/51737746e4914b24de048e36d3ac7e43f16155f5/pkg/reconciler/reconciler.go#L247) saying "an empty namespace may not be set when a resource name is provided" and loop is ending on first PR leaving other two PRs (concurrency_limit - 1 PRs) in queued state and as they are removed from semaphore they aren't no more in acquired list when next time ReconcileKind is called. This PR addresses that bug.


https://issues.redhat.com/browse/SRVKP-5897

savitaashture commented on test/gitea_test.go:
```diff
@@ -720,6 +720,79 @@ func verifyProvinance(t *testing.T, topts *tgitea.TestOpts, expectedOutput, cNam
+	verifyConcurrency(t, topts)
+}
+
+// TestGiteaGlobalAndLocalRepoCurrencyLimit verifies the concurrency_limit feature of the PipelineRun,
```
```suggestion
// TestGiteaGlobalAndLocalRepoConcurrencyLimit verifies the concurrency_limit feature of the PipelineRun,
```

enarha commented on pkg/reconciler/reconciler.go:
```diff
@@ -234,6 +234,13 @@ func (r *Reconciler) updatePipelineRunToInProgress(ctx context.Context, logger *
+		// in ReconcileKind this r.secretNS is not set and it is needed when git provider
+		// is not github.
+		r.secretNS = repo.GetNamespace()
+		if repo.Spec.GitProvider != nil && repo.Spec.GitProvider.Secret == nil && r.globalRepo.Spec.GitProvider != nil && r.globalRepo.Spec.GitProvider.Secret != nil {
```
We have the same code in reportFinalStatus. Can we have it in a single place? I mean if we have it in updatePipelineRunToInProgress, is it really needed in reportFinalStatus as well? Is it possible that reportFinalStatus is called without updatePipelineRunToInProgress being called earlier?

zakisk replied:
> Is it possible that reportFinalStatus is called without updatePipelineRunToInProgress being called earlier?

@enarha yes it is possible that `reportFinalStatus` will be called without `updatePipelineRunToInProgress` if `concurrency_limit` is not set and all `PipelineRuns` are provisioned in `STARTED` state by controller.

enarha replied:
Understood. Isn't there a common path (always executed) where we can move that code?

zakisk replied:
> Understood. Isn't there a common path (always executed) where we can move that code?

🤔 Yes, at the beginning of `ReconcileKind` it can be.

zakisk replied:
But we don't have `repo` reference object there in `ReconcileKind`, should we fetch repo object there?

zakisk replied:
> should we fetch repo object there?

@enarha if yes then  I will need to fetch repo in `ReconcileKind` and change `queue_pipelineruns` func signature to pass repo.

piyush-garg commented on pkg/reconciler/reconciler.go:
```diff
@@ -234,6 +234,13 @@ func (r *Reconciler) updatePipelineRunToInProgress(ctx context.Context, logger *
 	} else {
+		// in ReconcileKind this r.secretNS is not set and it is needed when git provider
+		// is not github.
+		r.secretNS = repo.GetNamespace()
```
Do we need to set this in reconciler field? can't we just pass the value while getting the secret, instead of setting it on reconciler?

zakisk replied:
@piyush-garg I think `r.secretNS` is the field which is meant for fetching secret for PipelineRun, if you see its uses it is used only two times once in [reportFinalStatus](https://github.com/openshift-pipelines/pipelines-as-code/blob/2d465fe57657b6ad04e409fa8c951f29c2e33c77/pkg/reconciler/reconciler.go#L169) and in [updatePipelineRunToInProgress](https://github.com/openshift-pipelines/pipelines-as-code/blob/2d465fe57657b6ad04e409fa8c951f29c2e33c77/pkg/reconciler/reconciler.go#L244).

piyush-garg replied:
but in case of global repo and all, setting this param can may be have adverse affect on next event etc, i would rather skip it for now like setting in reconciler

savitaashture replied:
I m not sure about the effect now but when i went through the code i see even r.secretNS is set either by `repo.GetNamespace()` or `r.secretNS = r.globalRepo.GetNamespace()` for globalRepo

@piyush-garg do you think it cause some issue :thinking:

chmouel replied:
I see the point of @piyush-garg even if i don't think it would cause problem but i think @zakisk you can remove that variable being on the struct and instead be scoped to the function that would be at least a saner code.

chmouel commented:
can you explain a bit more the problem statement of what you are solving, i am guessing myself and other reviewers can guess it from your PR but neither the SRVKP or the PR description describe how this bugs occurring and how to replicate it.

chmouel commented on test/gitea_test.go:
```diff
@@ -720,6 +720,79 @@ func verifyProvinance(t *testing.T, topts *tgitea.TestOpts, expectedOutput, cNam
+}
+
+func verifyConcurrency(t *testing.T, topts *tgitea.TestOpts) {
+	ctx := context.Background()
```
i think we want to move this helper method out of gitea_test.go to the pkg directory and mark it as t.Helper()

chmouel reviewed (commented):
pr is fine, just a few cleaning comment that would be needed....

zakisk commented:
> how this bugs occurring?

@chmouel when PRs are acquired in [queue_pipelinerun](https://github.com/openshift-pipelines/pipelines-as-code/blob/51737746e4914b24de048e36d3ac7e43f16155f5/pkg/reconciler/queue_pipelineruns.go#L54) from semaphore and [looped](https://github.com/openshift-pipelines/pipelines-as-code/blob/51737746e4914b24de048e36d3ac7e43f16155f5/pkg/reconciler/queue_pipelineruns.go#L59), (let assume that I got 5 PR and `concurrency_limit` is 3 so I got 3 PRs in [acquired list](https://github.com/openshift-pipelines/pipelines-as-code/blob/51737746e4914b24de048e36d3ac7e43f16155f5/pkg/reconciler/queue_pipelineruns.go#L54)), for first PipelineRun `r.secretNS` value is empty in the loop [at this point](https://github.com/openshift-pipelines/pipelines-as-code/blob/51737746e4914b24de048e36d3ac7e43f16155f5/pkg/reconciler/queue_pipelineruns.go#L66) when calling [updatePipelineRunToInProgress](https://github.com/openshift-pipelines/pipelines-as-code/blob/51737746e4914b24de048e36d3ac7e43f16155f5/pkg/reconciler/reconciler.go#L221) and updatePipelineRunToInProgress returning error from [this line](https://github.com/openshift-pipelines/pipelines-as-code/blob/51737746e4914b24de048e36d3ac7e43f16155f5/pkg/reconciler/reconciler.go#L247) saying `"an empty namespace may not be set when a resource name is provided"` and loop is ending on first PR leaving other two PRs (`concurrency_limit` - 1 PRs) in queued state and as they are removed from semaphore they aren't no more in acquired list when next time ReconcileKind is called.

zakisk commented:
> how to reproduce it?

I had encounter this while I was writing gitea tests for [SRVKP-5560](https://issues.redhat.com/browse/SRVKP-5560) ([PR link](https://github.com/openshift-pipelines/pipelines-as-code/pull/1745)), so if you run [tests](https://github.com/zakisk/pipelines-as-code/blob/51737746e4914b24de048e36d3ac7e43f16155f5/test/gitea_test.go#L642) of this PR without that `r.secretNS` assigment code, or any other existing concurrency test with `concurrency_limit` greater than 1, you will get that bug e.g. [TestGiteaConcurrencyExclusivenessMultiplePipelines](https://github.com/zakisk/pipelines-as-code/blob/51737746e4914b24de048e36d3ac7e43f16155f5/test/gitea_test.go#L172).

zakisk committed:
Fix empty secret namespace issue in reconciler

fixed empty secret namespace issue in reconcile when
git proivder is not github e.g. gitea, gitlab etc.
added E2E tests as well to confirm behaviour.

https://issues.redhat.com/browse/SRVKP-5897

Signed-off-by: Zaki Shaikh <zashaikh@redhat.com>

chmouel commented:
> > how this bugs occurring?
> 
> @chmouel when PRs are acquired in [queue_pipelinerun](https://github.com/openshift-pipelines/pipelines-as-code/blob/51737746e4914b24de048e36d3ac7e43f16155f5/pkg/reconciler/queue_pipelineruns.go#L54) from semaphore and [looped](https://github.com/openshift-pipelines/pipelines-as-code/blob/51737746e4914b24de048e36d3ac7e43f16155f5/pkg/reconciler/queue_pipelineruns.go#L59), (let assume that I got 5 PR and `concurrency_limit` is 3 so I got 3 PRs in [acquired list](https://github.com/openshift-pipelines/pipelines-as-code/blob/51737746e4914b24de048e36d3ac7e43f16155f5/pkg/reconciler/queue_pipelineruns.go#L54)), for first PipelineRun `r.secretNS` value is empty in the loop [at this point](https://github.com/openshift-pipelines/pipelines-as-code/blob/51737746e4914b24de048e36d3ac7e43f16155f5/pkg/reconciler/queue_pipelineruns.go#L66) when calling [updatePipelineRunToInProgress](https://github.com/openshift-pipelines/pipelines-as-code/blob/51737746e4914b24de048e36d3ac7e43f16155f5/pkg/reconciler/reconciler.go#L221) and updatePipelineRunToInProgress returning error from [this line](https://github.com/openshift-pipelines/pipelines-as-code/blob/51737746e4914b24de048e36d3ac7e43f16155f5/pkg/reconciler/reconciler.go#L247) saying `"an empty namespace may not be set when a resource name is provided"` and loop is ending on first PR leaving other two PRs (`concurrency_limit` - 1 PRs) in queued state and as they are removed from semaphore they aren't no more in acquired list when next time ReconcileKind is called.

great but can you add this to the PR description, the git commit and the SRVKP ? that will makes easier for everyone to know what the code was doing..

chmouel commented:
> > how to reproduce it?
> 
> I had encounter this while I was writing gitea tests for [SRVKP-5560](https://issues.redhat.com/browse/SRVKP-5560) ([PR link](https://github.com/openshift-pipelines/pipelines-as-code/pull/1745)), so if you run [tests](https://github.com/zakisk/pipelines-as-code/blob/51737746e4914b24de048e36d3ac7e43f16155f5/test/gitea_test.go#L642) of this PR without that `r.secretNS` assigment code, or any other existing concurrency test with `concurrency_limit` greater than 1, you will get that bug e.g. [TestGiteaConcurrencyExclusivenessMultiplePipelines](https://github.com/zakisk/pipelines-as-code/blob/51737746e4914b24de048e36d3ac7e43f16155f5/test/gitea_test.go#L172).

yeah just wanted you add this to the PR description to explain to the reviewer like me (or qe engineers) when looking at the PR how to test your PR easily without having to read the code.....

zakisk commented:
> great but can you add this to the PR description, the git commit and the SRVKP ? that will makes easier for everyone to know what the code was doing..

@chmouel oh, ok I will, I missed that.

zakisk commented:
> yeah just wanted you add this to the PR description to explain to the reviewer like me (or qe engineers) when looking at the PR how to test your PR easily without having to read the code.....

Added reproduction steps to [SRVKP](https://issues.redhat.com/browse/SRVKP-5897).

chmouel commented:
LGTM, good work @zakisk
